python/ray/data/dataset.py

```diff
@@ -602,6 +602,47 @@ def drop_columns(
             lambda batch: batch.drop(columns=cols), compute=compute, **ray_remote_args
         )
 
+    def select_columns(
+        self,
+        columns: List[str],
+        *,
+        compute: Union[str, ComputeStrategy] = None,
+        **ray_remote_args,
+    ) -> "Dataset[T]":
+        """Select one or more columns from the dataset.
+
+        Columns passed in will be de-duped since ArrowBlock and PandasBlock
+        `select` does not explicitly handle duplicated columns.
+
+        Examples:
+            >>> import ray
+            >>> # Create a dataset with 3 columns
+            >>> ds = ray.data.from_items([{"col1": i, "col2": i+1, "col3": i+2}
+            ...      for i in range(10)])
+            >>> # Select only "col1" and "col2" columns.
+            >>> ds = ds.select_columns(["col1", "col2"])
+            >>> ds
+            Dataset(num_blocks=10, num_rows=10, schema={col1: int64, col2: int64})
+
+
+        Time complexity: O(dataset size / parallelism)
+
+        Args:
+            columns: Names of the columns to select. Columns not included in this
+                will be filtered out.
+            compute: The compute strategy, either "tasks" (default) to use Ray
+                tasks, or ActorPoolStrategy(min, max) to use an autoscaling actor pool.
+            ray_remote_args: Additional resource requirements to request from
+                ray (e.g., num_gpus=1 to request GPUs for the map tasks).
+        """
+        # dedup the input columns used for selection
+        unique_columns = list(set(columns))
```

Comment:
Let's not do the deduplication. It will be another surprising different behavior between Ray and other systems (Arrow, Pandas, Spark, etc). Users usually don't add same columns multiple times. When they add, there's usually a valid reason for that. 

`select_columns(cols)` should just give users the exactly `cols` as they require.

Reply:
This seems not work well end to end. E.g. ds.select_columns(["x", "x"]).write_parquet("/tmp/pq"), and then user can no longer read it back due to duplicate column names.

Reply:
Usually users will rename columns if they are the same, so it's not a real issue. PyArrow has the API - https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.rename_columns .

Reply:
The issue is we don't have rename operator. If we do, it's actually not duplicate name anymore, just like "SELECT x as c1, x as c2", it actually has c1 and c2 as output (not duplicate x).

PyArrow cannot handle read back as well:
```
>>> ty = pq.read_table("/tmp/test.pq")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/anaconda3/lib/python3.9/site-packages/pyarrow/parquet/__init__.py", line 2827, in read_table
    return dataset.read(columns=columns, use_threads=use_threads,
  File "/home/ubuntu/anaconda3/lib/python3.9/site-packages/pyarrow/parquet/__init__.py", line 2473, in read
    table = self._dataset.to_table(
  File "pyarrow/_dataset.pyx", line 331, in pyarrow._dataset.Dataset.to_table
  File "pyarrow/_dataset.pyx", line 298, in pyarrow._dataset.Dataset.scanner
  File "pyarrow/_dataset.pyx", line 2361, in pyarrow._dataset.Scanner.from_dataset
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Multiple matches for FieldRef.Name(x) in x: int64
x: int64
```

Reply:
Users can use `map_batches` to rename columns with [Arrow](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.rename_columns) and [Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html) right? So don't have a first-class renaming API is not a blocker.